### ISSUE TITLE
Gracefully handle hyphens in hostnames

### ISSUE THREAD

**Issue by matosconsulting** [mitmproxy/mitmproxy#3885]
#### Problem Description
Attempting to set up a reverse proxy via web UI with the following string results in an "Invalid hostname" error.

reverse:http://api-.example.com/

#### Steps to reproduce the behavior through Web UI:
1. Log in to the mitmproxy web ui
2. Click options
3. Put "reverse:http://api-.example.com" in the mode box.

#### Steps to reproduce the behavior through Command Line:
1. Run this command: _mitmweb -m reverse:http://api-.example.com/_
2. See Response: _mitmweb: Invalid hostname: api-.example.com_

**Comment by mhils** [comment]
I guess we do some RFC 952 parsing here which mandates that parts of a hostname must not end with a hyphen (`api-` does end with a hyphen). Is there a convincing use case to disregard the standards here?

See also: https://stackoverflow.com/questions/2180465/can-domain-name-subdomains-have-an-underscore-in-it

**Comment by matosconsulting** [comment]
tldr: Despite the RFCs, these subdomains exist in the wild. Given the purpose of the tool I believe the mere existence of these subdomains is a compelling reason to allow for trailing hyphens in subdomains.

--

I think there's consensus that hostnames/subdomains _shouldn't_ have trailing hyphens, but acknowledgement that some registrars/DNS providers allow for it. In fact, the reason I opened this bug was because I was attempting to do some MITM logging for an IOT device which communicates to a hyphenated subdomain.

In fact, I just went in added some hyphenated A and C records on my domain, pointing to example.com.

```
> nslookup hyphenateda-.matosconsulting.com

Non-authoritative answer:
Name:    hyphenateda-.matosconsulting.com
Address:  93.184.216.34


> nslookup hyphenatedc-.matosconsulting.com

Non-authoritative answer:
Name:    example.com
Addresses:  2606:2800:220:1:248:1893:25c8:1946
          93.184.216.34
Aliases:  hyphenatedc-.matosconsulting.com
```

DNS lookups worked in Windows. Kali linux was spotty (ping worked, nslookup did not). I did some browser tests on trailing hyphen subdomains and found that while Edge puked on them (redirecting to a Bing search), Chrome and Firefox connected and rendered just fine (Windows, Kali, Android 10).

Discussion on another project: https://github.com/google/guava/issues/1065

**Comment by mhils** [comment]
Right. I just found that we also have precedent in mitmproxy to handle such cases more carefully:
https://github.com/mitmproxy/mitmproxy/issues/1529. I'd be happy to merge a PR that allows hyphens and adds a small test case. #1905 serves as a great template! 😃

**Comment by matosconsulting** [comment]
I will submit a PR tonight. Thanks!

**Comment by matosconsulting** [comment]
PR submitted with passing checks. This was my first time doing anything in Python and I might have gone slightly overboard with the additional checks...

**Comment by Prinzhorn** [comment]
@matosconsulting In the future, if you're using keywords such as "closes #xxx" or "fixes #xxx" in either your commit or the PR description it will automatically close the corresponding issue once the PR is merged.

https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
